### PR TITLE
Validate IAM

### PR DESCRIPTION
**Pull request by sean-smith**
Validates the role the user provided `ec2_iam_role`, if specified. 

### Testing
1. Run with `ec2_iam_role = cfncluster-old` set in config
   1. `cfncluster-old` is the role currently at https://cfncluster.readthedocs.io/en/latest/iam.html
```
IAM role error on user provided role cfncluster-old: action ec2:DescribeVolumes is implicitDeny
```
1. Run with `ec2_iam_role = cfncluster-new`
   1. `cfncluster-new` is cfncluster-old plus `ec2:DescribeVolumes` action
1. Run with `ec2_iam_role = admin`
   1. `admin` is just a role with an attached `AdministratorAccess` policy
1. Run with `ec2_iam_role = none`
   1. `none` is a role with no policy attached
```
IAM role error on user provided role none: action ec2:DescribeVolumes is implicitDeny
```
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/cfncluster/config_sanity.py [hunk 1]
```diff
@@ -33,6 +35,35 @@ def check_resource(region, aws_access_key_id, aws_secret_access_key, resource_ty
...
+                        (['dynamodb:PutItem', 'dynamodb:Query', 'dynamodb:GetItem', 'dynamodb:DeleteItem', 'dynamodb:DescribeTable'], "arn:aws:dynamodb:%s:%s:table/cfncluster-*" % (region, accountid)),
+                        (['sqs:ListQueues'], "*"),
+                        (['logs:*'], "arn:aws:logs:*:*:*")]
+
+            for actions, resource_arn in iam_policy:
+                response = iam.simulate_principal_policy(PolicySourceArn=arn, ActionNames=actions, ResourceArns=[resource_arn])
+                for decision in response.get("EvaluationResults"):
+                    if decision.get("EvalDecision") != "allowed":
+                        print("IAM role error on user provided role %s: action %s is %s" %
+                              (arn, decision.get("EvalActionName"), decision.get("EvalDecision")))
```
Comment (mauri-melato) [inline cli/cfncluster/config_sanity.py]: I think we should print the role name provided by the user (`resource_value`) and not the arn.

Comment (sean-smith) [reply cli/cfncluster/config_sanity.py]: changed.

**Comment by sean-smith** [comment]
This needs to be merged after boto3 #397

File: cli/cfncluster/cfnconfig.py [hunk 2]
```diff
@@ -228,7 +228,7 @@ def __init__(self, args):
                                       cwl_log_group=('CWLLogGroup',None),shared_dir=('SharedDir',None),tenancy=('Tenancy',None),
                                       ephemeral_kms_key_id=('EphemeralKMSKeyId',None), cluster_ready=('ClusterReadyScript','URL'),
                                       master_root_volume_size=('MasterRootVolumeSize',None),compute_root_volume_size=('ComputeRootVolumeSize',None),
-                                      base_os=('BaseOS',None),ec2_iam_role=('EC2IAMRoleName','EC2IAMRoleName'),extra_json=('ExtraJson',None),
+                                      base_os=('BaseOS',None),ec2_iam_role=('EC2IAMRoleName', 'EC2IAMRoleName'),extra_json=('ExtraJson',None),
```
Comment (mohanasudhan) [inline cli/cfncluster/cfnconfig.py]: remove space from this line or add space between (x, y) in the mapping, just to be consistent

File: cli/cfncluster/config_sanity.py [hunk 3]
```diff
@@ -33,6 +33,35 @@ def check_resource(region, aws_access_key_id, aws_secret_access_key, resource_ty
...
             print('Config sanity error: %s' % e.response.get('Error').get('Message'))
             sys.exit(1)
+    if resource_type == 'EC2IAMRoleName':
+        try:
+            iam = boto3.client('iam', region_name=region,
+                                aws_access_key_id=aws_access_key_id,
+                                aws_secret_access_key=aws_secret_access_key)
+
+            arn = iam.get_role(RoleName=resource_value).get('Role').get('Arn')
+            accountid = boto3.client('sts').get_caller_identity().get('Account')
```
Comment (mohanasudhan) [inline cli/cfncluster/config_sanity.py]: Make sure you add get_caller_identity permission in CfnclusterUserPolicy

File: cli/cfncluster/config_sanity.py [hunk 4]
```diff
@@ -33,6 +33,35 @@ def check_resource(region, aws_access_key_id, aws_secret_access_key, resource_ty
...
+                        (['dynamodb:ListTables'], "*"),
+                        (['sqs:SendMessage', 'sqs:ReceiveMessage', 'sqs:ChangeMessageVisibility', 'sqs:DeleteMessage', 'sqs:GetQueueUrl'], "arn:aws:sqs:%s:%s:cfncluster-*" % (region, accountid)),
+                        (['autoscaling:DescribeAutoScalingGroups', 'autoscaling:TerminateInstanceInAutoScalingGroup', 'autoscaling:SetDesiredCapacity'], "*"),
+                        (['cloudwatch:PutMetricData'], "*"),
+                        (['dynamodb:PutItem', 'dynamodb:Query', 'dynamodb:GetItem', 'dynamodb:DeleteItem', 'dynamodb:DescribeTable'], "arn:aws:dynamodb:%s:%s:table/cfncluster-*" % (region, accountid)),
+                        (['sqs:ListQueues'], "*"),
+                        (['logs:*'], "arn:aws:logs:*:*:*")]
+
+            for actions, resource_arn in iam_policy:
+                response = iam.simulate_principal_policy(PolicySourceArn=arn, ActionNames=actions, ResourceArns=[resource_arn])
```
Comment (mohanasudhan) [inline cli/cfncluster/config_sanity.py]: same here, iam::simulate_principal_policy api call is newly added. Please make sure CfnclusterUserPolicy is updated

File: cli/cfncluster/cfnconfig.py (hunk 2 above)
Comment (sean-smith) [reply cli/cfncluster/cfnconfig.py]: done.

File: cli/cfncluster/config_sanity.py (hunk 3 above)
Comment (sean-smith) [reply cli/cfncluster/config_sanity.py]: will update docs

File: cli/cfncluster/config_sanity.py (hunk 4 above)
Comment (sean-smith) [reply cli/cfncluster/config_sanity.py]: will update docs

**Pull request merged**